File: app/src/main/java/br/com/popularmovies/moviedetail/reviews/viewModel/MovieReviewViewModel.java
package br.com.popularmovies.moviedetail.reviews.viewModel;

import androidx.arch.core.util.Function;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import br.com.popularmovies.data.model.Resource;
import br.com.popularmovies.services.movieService.response.MovieReviews;
import br.com.popularmovies.services.movieService.source.MovieRepository;

public class MovieReviewViewModel extends ViewModel {
    private final LiveData<Resource<MovieReviews>> reviews;
    private final MutableLiveData<Integer> movieId = new MutableLiveData<>();

    public MovieReviewViewModel(final MovieRepository mMovieRepository, int movieId) {
        reviews = Transformations.switchMap(this.movieId, new Function<Integer, LiveData<Resource<MovieReviews>>>() {
            @Override
            public LiveData<Resource<MovieReviews>> apply(Integer input) {
                return input != null ? mMovieRepository.getMovieReviews(input) : null;
            }
        });
        this.movieId.setValue(movieId);
    }

    public LiveData<Resource<MovieReviews>> getReviews() {
        return reviews;
    }

    public void tryAgain() {
        movieId.setValue(movieId.getValue());
    }
}

File: app/src/main/java/br/com/popularmovies/moviedetail/reviews/viewModel/MovieReviewViewModel.kt
package br.com.popularmovies.moviedetail.reviews.viewModel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.com.popularmovies.common.models.base.NetworkError
import br.com.popularmovies.common.models.base.Result
import br.com.popularmovies.entities.movie.MovieReview
import br.com.popularmovies.usecases.movies.reviews.GetMovieReviewsUseCase
import com.squareup.inject.assisted.Assisted
import com.squareup.inject.assisted.AssistedInject
import kotlinx.coroutines.launch

class MovieReviewViewModel @AssistedInject constructor(
    val getMovieReviewsUseCase: GetMovieReviewsUseCase,
    @Assisted val movieId: Long
) : ViewModel() {

    @AssistedInject.Factory
    interface Factory {
        fun create(movieId: Long): MovieReviewViewModel
    }

    val loading = MutableLiveData<Boolean>()

    private val _error = MutableLiveData<NetworkError>()
    val error: LiveData<NetworkError>
        get() = _error
    private val _reviews = MutableLiveData<List<MovieReview>>()
    val reviews: LiveData<List<MovieReview>>
        get() = _reviews

    init {
        getReviews()
    }

    fun getReviews() {
        viewModelScope.launch {
            showLoading(true)
            val params = GetMovieReviewsUseCase.Params(movieId)
            when (val result = getMovieReviewsUseCase.build(params)) {
                is Result.Success -> _reviews.value = result.data
                is Result.Error -> _error.value = result.error
            }
        }
    }

    fun showLoading(value: Boolean) {
        loading.value = value
    }

    fun tryAgain() {
        getReviews()
    }
}
